### ISSUE TITLE
On trigonometric altitude calculations

### ISSUE THREAD

**Issue by bmorris3** [astropy/astroplan#179]
TL;DR: maybe it's no longer "premature optimization" to optimize rise/set/transit time calculations within astroplan. Let's talk about it!

I'd like to start a discussion about using trigonometric solutions to rise/set/transit time calculations, rather than using an iterative grid method as is currently the default behavior. Posting the issue was prompted by a conversation with @StuartLittlefair who noted: 

> One thing I've noticed that could be improved is the calculation of rise/set and transit times. At the moment you grid over 24 hours to search for the appropriate event, but it would be possible to use spherical trig to hone in on the event time and grid over a smaller time range, thus increasing the accuracy (at the moment the times of rise/set/midnight can change by ~minutes depending on the input time).

This was a point of discussion when I started writing the guts of astroplan last summer. There is currently a [private method on the Observer object, `Observer._altitude_trig`](https://github.com/astropy/astroplan/blob/master/astroplan/observer.py#L638) that does the spherical trig to calculate the altitude of a target as a function of time, which can be used to calculate rise/set/transit time. In a simple benchmark test that I did, I think theres a factor of ~3x speedup using the spherical trig method instead of using the full astropy machinery.

The reasons why the trig solution is not the default behavior are summarized by an old comment from
@eteq, [which I encourage you to read over here](https://github.com/astropy/astroplan/pull/14#issuecomment-117430584). In summary, Erik expects that the trigonometric calculation probably gets to the correct answer to within ~minutes, and the astropy solution probably gives you a solution to within a few ~seconds. He suspects that it will be least confusing for developers and users if the altitude calculations are done consistently at high-precision, so we're using astropy everywhere even though it's slower.

@StuartLittlefair is interested in testing hybrid methods. Maybe one place to start is to go ahead and implement the hybrid method and benchmark it against the current default. It would also be interesting to see how wrong the trigonometric solutions are. You could generate a list of targets that span the sky, and calculate the difference between the rise/set times for a bunch of observers, using the current default behavior and the trig solution. Then at least we can quantify what the trade-offs are. 

I still think there's a place for a keyword argument toggle on the constraints that switches to a fast, low-precision trig method. While it's important to have the highest possible precision when you call `Observer.altaz`, I think that no one expects constraints to be computed with _O_(second) precision, because in practice no observations can ever be planned that precisely.

cc @ejeschke

**Comment by StuartLittlefair** [comment]
To put some flesh on this discussion, I implemented a simple routine to calculate the rise/set time using spherical trig. Obviously it does not account for the atmosphere. It currently does not handle circumpolar and never rise objects, which will slow it down marginally.

I tested the difference between the two routines by randomly sampling a sphere for targets and then throwing away objects that never rose or set. The routine's answers differ from the current implementation by ~3 minutes; the distribution of the discrepancy is approximately uniform between 0 and 3.5 minutes. It is roughly 5 times faster than the current implementation, though much of this time is spent ensuring the input coordinates are in the ICRS frame. It's 20-30 times faster if the coordinates are already in the ICRS frame.

I think for observers/schedulers with large numbers of ```Constraints``` to compute, this speedup would be worthwhile, although obviously it comes at a significant accuracy penalty. 

For info; the kind of 'hybrid' methods I was thinking of were:

- use the trigonometric answers as a starting guess for a binary chop algorithm
- use the current approach (gridding times), but use the trigonometric answers to narrow the current grid from a 24 (N=150) hour grid to a ~1 (N<15) hour grid which could be both faster and more accurate.

However, I am wary of premature optimisation and the current implementation is both fast enough and accurate enough for my own use case, so I wonder if this should wait until demand arises?

**Comment by StuartLittlefair** [comment]
To revisit this, my PhD student is now finding that evaluating ```Constraint```s is sufficiently slow as to make each scheduling step too slow for our requirements; we ideally want to keep ```Constraint``` evaluation for our scheduling step to <10s even with ~100 targets.

If people approve (@bmorris3, @eteq) I'll begin work on two PRs; one to add trigonometric rise_set and meridian_transit functions and a second one to allow these to be used in constraint evaluations if required.

**Comment by bmorris3** [comment]
This improvement is implemented in #327, but deprecated in favor of #424.
